sdk/client/src/stream/streaming_client.rs

```diff
@@ -146,17 +152,40 @@ impl StreamingClient {
     pub async fn send_subscription(
         &mut self,
         request: StreamMethodRequest,
-    ) -> SubscriptionStreamResult {
-        let subscription_stream = self.get_and_register_id().await?;
+    ) -> StreamResult<SubscriptionStream> {
+        let mut subscription_stream = self.get_and_register_id().await?;
         let res = self
             .client
             .write()
             .await
-            .send_method_request(request, Some(subscription_stream.id.clone()))
+            .send_method_request(request, Some(subscription_stream.id().clone()))
             .await;
 
-        self.maybe_clear_subscription(&subscription_stream.id, res)
-            .await?;
+        let id = match res {
+            Ok(id) => id,
+            Err(e) => {
+                self.clear_subscription(&subscription_stream.id()).await;
```

Comment:
This looks quite error prone, requiring that we explicitly call clear_subscription on each of the return paths. Is there a way we could use RAII to avoid this?